crates/extensions/c8y_mapper_ext/src/operations/mod.rs

```diff
@@ -12,48 +12,313 @@
+    /// operations to be processed. Messages not related to operations are ignored.
+    ///
+    /// `entity` needs to be the same entity as in `message`.
+    ///
+    /// When a message with a new id is handled, a task will be spawned that processes this
+    /// operation. For the operation to be completed, subsequent messages with the same command id
+    /// need to be handled as well.
+    ///
+    /// When an operation terminates (successfully or unsuccessfully), an MQTT operation clearing
+    /// message will be published automatically, which does not need to be handled.
+    pub async fn handle(&mut self, entity: EntityTarget, message: MqttMessage) {
+        let Ok((_, channel)) = self.context.mqtt_schema.entity_channel_of(&message.topic) else {
+            return;
+        };
 
-    // used to automatically remove the temporary file after operation is finished
-    pub file_dir: tempfile::TempDir,
+        let Channel::Command { operation, cmd_id } = channel else {
+            return;
+        };
 
-    // TODO: remove this message field since command is available
-    // the message that triggered the operation
-    pub message: MqttMessage,
+        let message = OperationMessage {
+            operation,
+            entity,
+            cmd_id: cmd_id.into(),
+            message,
+        };
 
-    pub entity_topic_id: EntityTopicId,
+        let topic: Arc<str> = message.message.topic.name.clone().into();
+        let running_operation = {
+            let op = self.running_operations.get(&topic);
+
+            if let Some(running_operation) = op {
+                // task already terminated
+                if running_operation.tx.send(message).is_err() {
+                    let running_operation = self.running_operations.remove(&topic).unwrap();
+                    Some(running_operation)
+                } else {
+                    None
+                }
+            } else {
+                let running_operation = RunningOperation::spawn(message, Arc::clone(&self.context));
+
+                self.running_operations.insert(topic, running_operation);
+                None
+            }
+        };
 
-    pub command: GenericCommandState,
+        if let Some(running_operation) = running_operation {
+            let join_result = running_operation.handle.await;
+            error!("handle: {join_result:?}");
+        }
+    }
 }
 
-/// Used to denote as type of what operation was the file downloaded from the FTS.
+pub struct RunningOperation {
+    handle: tokio::task::JoinHandle<()>,
+    tx: tokio::sync::mpsc::UnboundedSender<OperationMessage>,
+}
+
+impl RunningOperation {
+    /// Spawns a task that handles the operation.
+    ///
+    /// The task handles a single operation with a given command id, and via a channel it receives
+    /// operation state changes (if any) to drive an operation to completion.
+    fn spawn(message: OperationMessage, context: Arc<OperationContext>) -> Self {
```

Comment:
Okay this design is correct. I'm not fully happy with it, though. That said, I acknowledge that the complexity is nicely grouped in one place and that all the operation *managers* (i.e. the misc `context.handle_XXX`) are fully unaware of it.

=> I will take more time to clarify my thoughts on that.

Reply:
> => I will take more time to clarify my thoughts on that.

Here is my point of view about this design:

Pros:

- The complexity is located in one place. Two to be precise:
  - the `RunningOperation::spawn()` function that spawn a running operation for a specific request
  - the `OperationHandler::handle()` method that manages the running operations and dispatches status messages up to completion.
- The operation handlers are focused on the task to be done
  - the sequence of steps to handle the outcome of an operation is written in plain async/await code
  - there is no more state cache between steps
  - see `OperationContext::handle_config_snapshot_state_change` for an example
  - this  alone is a reason to merge this PR

Cons:

- The `RunningOperation::spawn()` function  mixes low level aspects with high level ones.
  - tasks, join handles and channels on one side 
  - tedge operations and log uploads on the other side
  - plus a specific behavior on command clearing messages
- The `RunningOperation::spawn()` function is akin to an `Actor::run()` method
  - consuming messages in a loop up to completion,
  - but mixed with message delivery aspects 
- Each `OperationContext::handle_*()` method is also akin to an `Actor` message handler